### PR TITLE
Puppet 4 Katello Integration Blog Post.

### PR DESCRIPTION
**Pull request by zjhuntin**
So far I have an outline to what I want this Blog post to look like. Putting this up to share it with my team.

### CODE REVIEW

File: _posts/2016-08-28-katello-puppet4-integration.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+- katello
+- puppet 4
+excerpt: |
+    Puppet 4 integration is coming to Katello and we wanted to share with you
+    how it works!
+---
+
+### Puppet 4 Integration with Katello
+
+Recently Foreman has grown to work with Puppet 4. Katello is following suite and will soon be ready to use Puppet 4. In this post we will explain how to create new Katello environments Puppet 4 ready, and upgrade exisiting systems.
```
Comment (ehelms) [inline _posts/2016-08-28-katello-puppet4-integration.md]: We are ready for Puppet 4! Maybe indicate that by changing the second sentence to note that it is in nightly and will be included in 3.2 ?

File: _posts/2016-08-28-katello-puppet4-integration.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+There are some important things to know about environments with Katello and Puppet 4:
+
+Puppet 4 uses a Java Virtual Machine and requires more memory than Puppet 3 had before. This means that systems acting as Katello servers with Puppet 4 need a larger amount of memory. At a minimum we are requiring 8 GB of memory, and reccomending 16GB (<-- need a fact check on this). 
+
+Puppet 4 masters can still provision computers with Puppet 3 with [some configuration](https://docs.puppet.com/puppetserver/latest/compatibility_with_puppet_agent.html). This can help with systems that are migrating over to Puppet 4. You can start with upgrading your masters to Puppet 4 and then gradually upgrading you agents.
```
Comment (ehelms) [inline _posts/2016-08-28-katello-puppet4-integration.md]: s/computers/hosts

File: _posts/2016-08-28-katello-puppet4-integration.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+The --upgrade-puppet command will restart the services, but you can make sure they are running with `hammer ping`. If you run puppet --version on the server, it should show `4.6.0` or a smilar 
+
+### Contributors
+
+I would like to thank everyone on the Katello team for putting in the effort for making this happen. As well as Foreman, the community  that puts so much effort into these great projects!
+
+### More Information
+
+Here are some good places to find more information about Puppet4 and using it with Katello.
```
Comment (ehelms) [inline _posts/2016-08-28-katello-puppet4-integration.md]: A link to the demo @beav did would be a good one to include.

File: _posts/2016-08-28-katello-puppet4-integration.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+* After run:
+
+```bash
+vagrant up example-capsule-p4
+```
+
+#### Devlopment environment
+
+### Details on how to upgrade both Katello servers, capsules, and hosts to use puppet 4
```
Comment (ehelms) [inline _posts/2016-08-28-katello-puppet4-integration.md]: I'd consider this a higher interest item and would move it further up the blog post.

File: _posts/2016-08-28-katello-puppet4-integration.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+We have a [great video](https://youtu.be/GFNPHypFhl4?t=2545) and [excellent documentation](http://www.katello.org/docs/nightly/upgrade/puppet.html) on how to upgrade from Puppet 3 to 4.
+
+To upgrade an exisisting Katello instance to Puppet 4 it is straight forward:
+
+Before starting backup or take snapshot the Katello server.
```
Comment (ehelms) [inline _posts/2016-08-28-katello-puppet4-integration.md]: s/take snapshot/take a snapshot of

Comment (ehelms) [reply _posts/2016-08-28-katello-puppet4-integration.md]: Maybe reword as well to "Before starting, we recommend creating a backup or taking a snapshot...`

File: _posts/2016-08-28-katello-puppet4-integration.md (hunk 3 above)
Comment (ehelms) [reply _posts/2016-08-28-katello-puppet4-integration.md]: Oh I see it up above, maybe include it here too incase it was missed.

**Comment by ehelms** [comment]
I think it would be worth talking briefly about how Katello puts puppet content into the defualt locations for puppet 3 or puppet 4. And some mention of allowing for mixed infrastructure of Puppet 3 and Puppet 4 servers for those who want to migrate gradually.

**Comment by zjhuntin** [comment]
@ehelms I went ahead and added in things for some of your comments. It's not totally finished yet, but do you mind taking another look at it?

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+katello-service stop # This will stop all services.
+foreman-installer --upgrade-puppet # This performs the upgrade.
+```
+
+The --upgrade-puppet command will restart the services, but you can check that they are running with `hammer ping`. If you run puppet --version on the server, it should show `4.x.x`. 
+
+### How to install Katello with Puppet 4
+
+#### Production Environments
+If you want to create a new Katello server configured with Puppet 4, the easiest way is to spin up a Virtual Machine with [Forklift](https://github.com/Katello/forklift).
```
Comment (ehelms) [inline _posts/2016-10-19-katello-puppet4-integration.md]: This may be better to emphasize for users who want to test out a production environment with Puppet 4 as we typically don't recommend that users use the Vagrant setup for an installation they are planning to put into real production.

**Comment by ehelms** [comment]
@stbenjam mind giving this a review as well?

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 7]
```diff
@@ -0,0 +1,132 @@
+---
+layout: post
+title: Katello Puppet4 integration
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: Space between Puppet and 4

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+date: 2016-10-19
+author: Zach Huntington-Meath
+tags:
+- katello
+- puppet 4
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: I don't think tags can have a space. I'd just tag it as "puppet" since there's a bunch of other posts w/ that tag.

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+excerpt: |
+    Puppet 4 integration is coming to Katello and we wanted to share with you how it works!
+---
+
+### Puppet 4 Integration with Katello
+
+Recently, Foreman and Puppet 4 became compatible. Now, Katello is ready to work with Puppet 4. It is available in Katello nightly and will be released in version 3.2.  In this post, we will explain how to create new Katello instances that are Puppet 4 ready, and upgrade existing systems.
+
+There are some important things to know about using Katello with Puppet 4:
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: I think the items below this should be bullet points (i.e. prefixed `-`)

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+
+Puppet 4 uses the Java Virtual Machine and requires more memory than Puppet 3 did. This means that systems acting as Katello servers with Puppet 4 need a larger amount of memory. At a minimum, we are requiring 8 GB of memory, and recommending 16GB (<-- need a fact check on this.)
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: I've been actively using 8GB P4 vms and seems OK.  Not sure about 16, I'd say 12 is a good recommendation.

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+You can use this matrix as a quick reference of what can work together:
+
+| | Puppet 3 client | Puppet 4 client
+| :--- | :---: | :---: 
+| **Puppet 4 server** | Compatible | Compatible
+| **Puppet 3 server** | Compatible | Not compatible
+
+Katello will by default install modules to the normal file paths for the version of Puppet you are using. 
+Puppet 3 will install modules to `/etc/puppet/environments`
+Puppet 4 will install modules to `/etc/puppetlabs/code/environments`
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: Seems like these should be bullet points

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+
+### How to upgrade both Katello servers, capsules, and hosts to use Puppet 4
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: There's 3 things here so shouldn't say 'both'

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+We have a [great video](https://youtu.be/GFNPHypFhl4?t=2545) and [excellent documentation](http://www.katello.org/docs/nightly/upgrade/puppet.html) on how to upgrade from Puppet 3 to 4.
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: I'd link to 3.2 docs, IMHO nightly is mostly for dev consumption

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+
+To upgrade an existing Katello instance to Puppet 4:
+
+Before starting, we recommend creating a backup or a snapshot of your system.
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: Do we have docs on how to backup to link to? If not, I'd mention something like "...with katello-backup."

File: _posts/2016-10-19-katello-puppet4-integration.md (hunk 6 above)
Comment (stbenjam) [reply _posts/2016-10-19-katello-puppet4-integration.md]: :+1: I'd prefer to just link to the install docs on katello.org here as the main installation method. And we need to fix them so they should add an option for users who want to use Puppet 3 vs. 4.  

Or, just start listing Puppet 4 as the default and note they can still use Puppet 3 if they prefer.

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+* In your boxes.yml file, you can write:
+```yaml
+example-capsule-p4:
+  box: centos7
+  ansible:
+    playbook: 'playbooks/capsule-dev.yml'
+    options: --puppet-four
+    group: 'capsule'
+    server: 'example-katello-p4'
+```
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: Note: If https://github.com/Katello/forklift/pull/279 is merged, we should change this to ansible

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+* After run:
+
+```bash
+vagrant up example-capsule-p4
+```
+
+* Once installation completes:
+
+```bash
+vagr
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: What is supposed to be here?

File: _posts/2016-10-19-katello-puppet4-integration.md [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+
+### More Information
+
+Here are some good places to find more information about Puppet4 and using it with Katello:
+
+[Puppet 4 Documentation](https://docs.puppet.com/puppet/4.5/reference/)
+
+[Katello Puppet 4 Documentation](http://www.katello.org/docs/nightly/upgrade/puppet.html)
+
+[Katello repository](https://github.com/Katello/katello)
```
Comment (stbenjam) [inline _posts/2016-10-19-katello-puppet4-integration.md]: I'd remove this, not sure it's super relevant to the post as it's not referenced anywhere

**Review by stbenjam — CHANGES_REQUESTED** [review]
This is great, thanks @zjhuntin! Just a few minor comments!

File: _posts/2016-10-19-katello-puppet4-integration.md (hunk 16 above)
Comment (zjhuntin) [reply _posts/2016-10-19-katello-puppet4-integration.md]: This was an incomplete part, sorry about that!

**Comment by zjhuntin** [comment]
@ehelms @stbenjam thank you both so much for the comments! I'll get on these.

**Comment by zjhuntin** [comment]
@ehelms @stbenjam Updated as per comments. Also added some more information to getting the development environment running properly.

File: _posts/2016-10-19-puppet4-integration.md [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+
+### How to upgrade Katello servers, capsules, and hosts to use Puppet 4
+
+We have a [great video](https://youtu.be/GFNPHypFhl4?t=2545) and [excellent documentation](http://www.katello.org/docs/3.2/upgrade/puppet.html) on how to upgrade from Puppet 3 to 4.
+
+To upgrade an existing Katello instance to Puppet 4:
+
+Before starting, we recommend creating a [backup](http://www.katello.org/docs/3.2/user_guide/backup/) or a snapshot of your system.
+
+Then run:
```
Comment (stbenjam) [inline _posts/2016-10-19-puppet4-integration.md]: Hm, I just noticed that we're missing a step here, the user has to make the packages available by enabling the Puppet 4 repo instead of Puppet 3. 

See "Upgrading" section of http://www.katello.org/docs/3.2/upgrade/puppet.html

File: _posts/2016-10-19-puppet4-integration.md [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+
+```bash
+git clone https://github.com/katello/forklift.git
+cd forklift
+vagrant up centos7-katello-p4
+```
+
+This will setup a virtual environment with Katello and Puppet 4 enabled.
+
+If you want to install a real production server with Puppet 4 we reccommend following the instructions listed in our [documentation](http://www.katello.org/docs/3.2/installation/index.html).
```
Comment (stbenjam) [inline _posts/2016-10-19-puppet4-integration.md]: I would move this to the top of this section, since that's what most users would want to do

File: _posts/2016-10-19-puppet4-integration.md [hunk 20]
```diff
@@ -0,0 +1,148 @@
...
+rake webpack:compile
+rails s
+```
+
+- Now you have a functioning development environment!
+
+
+#### Making Capsules
+
+If you wanted to create a capsule with Forklift, you can follow these steps:
```
Comment (stbenjam) [inline _posts/2016-10-19-puppet4-integration.md]: I'd add a link like the production section to the official docs

**Review by stbenjam — CHANGES_REQUESTED** [review]
Looks great!

**Comment by zjhuntin** [comment]
@stbenjam Okay, made those changes!

**Review by stbenjam — APPROVED** [review]

**Comment by stbenjam** [comment]
Awesome, looks great to me, if this is ready to go remove [WIP] from the subject

**Review by ehelms — APPROVED** [review]

**Comment by ehelms** [comment]
I think this is ready to be merged.

**Comment by stbenjam** [comment]
[WIP] still needs removing from the commit message, too please

**Comment by zjhuntin** [comment]
@stbenjam @ehelms Updated, I think it's ready to go!

File: _posts/2016-10-21-puppet4-integration.md [hunk 21]
```diff
@@ -0,0 +1,153 @@
...
+You will need to give your system access to the new Puppet packages. To do this follow the instructions [here](https://docs.puppet.com/puppet/4.6/reference/puppet_collections.html).
+
+Then run:
+
+```bash
+katello-service stop # This will stop all services.
+foreman-installer --upgrade-puppet # This performs the upgrade.
+```
+
+The --upgrade-puppet command will restart the services, but you can check that they are running with `hammer ping`. If you run puppet --version on the server, it should show `4.x.x`.
```
Comment (stbenjam) [inline _posts/2016-10-21-puppet4-integration.md]: Can you surround all the `--XXXXXX` commands (outside of code blocks, I think these are the only ones) with backticks? Otherwise markdown combines them into an em dash:

emdash

File: _posts/2016-10-21-puppet4-integration.md [hunk 22]
```diff
@@ -0,0 +1,153 @@
...
+- Puppet 4 uses the Java Virtual Machine and requires more memory than Puppet 3 did. This means that systems acting as Katello servers with Puppet 4 need a larger amount of memory. At a minimum, we are requiring 8 GB of memory, and recommending 12GB
+
+- Puppet 4 masters can still provision hosts with Puppet 3 with [some configuration](https://docs.puppet.com/puppetserver/latest/compatibility_with_puppet_agent.html). This can help with systems that are migrating over to Puppet 4. You can start with upgrading your masters to Puppet 4 and then gradually upgrade your agents.
+
+- You can use this matrix as a quick reference of what can work together:
+
+| | Puppet 3 client | Puppet 4 client
+| :--- | :---: | :---: 
+| **Puppet 4 server** | Compatible | Compatible
+| **Puppet 3 server** | Compatible | Not compatible
```
Comment (stbenjam) [inline _posts/2016-10-21-puppet4-integration.md]: The table looks a little funny on the blog post, is there some setting we could set so the Markdown renders table borders?

table

**Review by stbenjam — CHANGES_REQUESTED** [review]
Sorry, I hate to drag this out but wanted to make sure it rendered OK on the actual web site before merging, there's a few weird quirks if you wouldn't mind taking a look, and then I'll merge.

File: _posts/2016-10-21-puppet4-integration.md (hunk 21 above)
Comment (zjhuntin) [reply _posts/2016-10-21-puppet4-integration.md]: Will do!

File: _posts/2016-10-21-puppet4-integration.md (hunk 22 above)
Comment (zjhuntin) [reply _posts/2016-10-21-puppet4-integration.md]: So I thought that the Markdown itself would render them naturally, but I'm guessing its the way that we style markdown on theforeman.org. I'll give it a shot though and see if I can find a way to make it look better.

**Comment by zjhuntin** [comment]
@stbenjam I found a way of making the table with borders. I had to make it with html, but I think that it does look better.

**Comment by stbenjam** [comment]
Thanks, that is _much_ better

**Review by stbenjam — APPROVED** [review]

**Comment by stbenjam** [comment]
Merged as a84eb517b9a59a92e5c74af18e876cb523a3ed00, I made a small change to update the date to today.

Thanks everyone!
